README.md
# Flutter 整理

> [一些 UI 搜集整理](./example_normal/iamfine/flutter_ui/screenshots.md)

> [一些 APP 搜集整理](./example_normal/iamfine/flutter_app/screenshots.md)

> [一款基于 RethinkDB 的聊天 APP](./example_normal/iamfine/flutter_app/flutter_chat_app/README.md)

## Web

web 产物在`ts-pages`分支可查看。
[查看运行效果](https://miaopasiflutter.github.io/AwesomeProject/#/index)

## Android

![one_download](./screenshots/app_download.jpg)
[JD-Android-V0.1 下载地址](https://www.pgyer.com/1lJA)

## iOS

无苹果开发者账号, 暂时无法打包 APP, 请亲自进行编译并在模拟器或手机上运行


example_normal/CHANGELOG.md
### ### 2023-11-22

> [fix]:This application cannot tree shake icons fonts. It has non-constant instances of IconData at the following locations

> > 解决方案：打包时加上 --no-tree-shake-icons

```
This application cannot tree shake icons fonts. It has non-constant instances of IconData at the following locations:
  -
      file:///Users/zhuyuhui/Documents/Github/MiaoPaSiFlutter/AwesomeProject/example_normal/iamfine/flutter_app/flutter_to_do_app/lib/flutter_to_do_ap
      p/pages/home/widgets/addDialog.dart:117:41
  -
      file:///Users/zhuyuhui/Documents/Github/MiaoPaSiFlutter/AwesomeProject/example_normal/iamfine/flutter_app/flutter_to_do_app/lib/flutter_to_do_ap
      p/pages/home/widgets/taskCard.dart:64:17
  -
      file:///Users/zhuyuhui/Documents/Github/MiaoPaSiFlutter/AwesomeProject/example_normal/iamfine/flutter_app/flutter_to_do_app/lib/flutter_to_do_ap
      p/pages/detail/detail.dart:56:27
Target web_release_bundle failed: Exception: Avoid non-constant invocations of IconData or try to build again with --no-tree-shake-icons.
```

### 2023-10-23

使用 Firebase 通过电子邮件登录和注销。

[将 Firebase 添加到您的 Flutter 应用](https://console.firebase.google.com/project/awesomeproject-flutter/overview?hl=zh-cn)
具体操作：
1、[安装 Firebase CLI](https://firebase.google.com/docs/cli?hl=zh&authuser=0)
➜ npm install -g firebase-tools
2、使用您的 Google 帐号登录 Firebase
➜ firebase login
3、安装并运行 FlutterFire CLI
➜ dart pub global activate flutterfire_cli
4、在您的 Flutter 项目的根目录下，运行以下命令：
➜ flutterfire configure --project=awesomeproject-flutter
这会自动向 Firebase 注册您的每个平台应用，并向您的 Flutter 项目添加 lib/firebase_options.dart 配置文件。

### 2023-10-22

1、创建 flutter_mitchkoko 模块儿
sh hzy_rob_project.sh -n flutter_mitchkoko -l pro -t package

### 2023-10-21

1、创建 flutter_books 模块儿
sh hzy_rob_project.sh -n flutter_books -l pro -t package

[Panda 看书](https://juejin.cn/post/6844903907110420488)
[追书神器 API 整理合集一](https://juejin.cn/post/6844903480939773966)

### 2023-10-20

#### 1、使用[GitHub Pages](https://zhuanlan.zhihu.com/p/647963756) 预览编译出的 web 产物

新建`ts-pages`分支，将编译出的 web 产物放置到该分支下。

#### 2、添加一键上传脚本文件

python3.10 upload.py --type='apk' --upload='apk'

### 2023-10-19

#### 1、编译 web

flutter build web --web-renderer html
问题:[创建 flutter web 的构建版本时显示一个空页面](https://www.saoniuhuo.com/question/detail-2518136.html)

```
描述：在localhost中运行flutter run -d chrome时工作正常，但是当我准备发布它时，使用命令flutter build web，它显示一个空白页面，并通过控制台抛出一些一般性错误。

解决：
在/build/web index.html中，需要将href="/”改为href="./”，以便它最终可以工作，似乎href="”也可以工作
```

### 2023-10-18

#### 1、[添加 web 支持](https://docs.flutter.dev/platform-integration/web/building)

This application is not configured to build on the web.
To add web support to a project, run `flutter create .`.

运行 `flutter create .` 会创建 web 文件夹
运行 `flutter run -d chrome` 运行到 Chrome

#### 2、[ Flutter Web 预览时白屏解决方法 ](https://www.cnblogs.com/letleon/p/16869746.html)

原因是因为运行 flutter run 是自动选择渲染器

--- 桌面端 WEB 浏览器默认使用 CanvasKit 渲染器

--- 移动端 WEB 浏览器默认使用 HTML 渲染器

问题就出在了 CanvasKit 渲染器,他需要下载一些库的支持，

解决方法：

```javascript：
1：挂梯子
2：改渲染器
　　2.1：在命令行中指定渲染器（每次都得敲命令,run 和 build 都可用）
　　　　　　HTML渲染器
　　　　　　　　flutter run -d chrome --web-renderer html

　　　　　　CANVASKIT渲染器（默认 flutter run -d chrome 就是这个）
　　　　　　　　flutter build web --web-renderer canvaskit

　　2.2:在WEB index.html 中设置渲染器（只需要修改一次）
　　　　　　在 web/index.html 文件中，main.dart.js 之前加入一个 <script>
复制代码

<!-- 渲染模式 -->
<script type="text/javascript">
   let useHtml = true;//你懂的
   if(useHtml) {
     window.flutterWebRenderer = "html";
   } else {
     window.flutterWebRenderer = "canvaskit";
   }
 </script>
<!--  在main.dart.js之前 -->
<!--  <script src="main.dart.js" type="application/javascript"></script> -->
```

### 2023-10-17

1、[使用 flutter_launcher_icons 修改程序的图标](https://github.com/fluttercommunity/flutter_launcher_icons/)

# 一个软件包，可简化更新 Flutter 应用程序启动器图标的任务。

flutter_launcher_icons: ^0.13.1

2、[使用 package_rename 修改包名。](https://github.com/OutdatedGuy/package_rename/tree/main)
package_rename: ^1.3.3

### 2023-10-16

1、创建 extended_sliver_demo 模块儿
sh hzy_rob_project.sh -n extended_sliver_demo -l pro -t package

2、创建 scrollerview_demo 模块儿
sh hzy_rob_project.sh -n scrollerview_demo -l pro -t package

3、创建 animation_demo 模块儿
sh hzy_rob_project.sh -n animation_demo -l pro -t package

4、创建 canvas_demo 模块儿
sh hzy_rob_project.sh -n canvas_demo -l pro -t package

5、创建 common_demo 模块儿
sh hzy_rob_project.sh -n common_demo -l pro -t package

### 2023-10-15

1、创建 京东 模块儿
sh hzy_rob_project.sh -n flutter_jd -l pro -t package

2、添加京东登录页面

### 2023-10-14

1、补充自动化脚本

### 2023-10-13

Xcode14.2
Flutter SDK:3.13.7
1、固定 pubspec.yaml 中三方库的版本号

### 2023-10-12

项目初始化
Xcode13.2
Flutter SDK:3.10.0


example_normal/upload.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# @Time    : 2020-08-04 15:19
# @Author  : fgyong 简书:_兜兜转转_  https://www.jianshu.com/u/6d1254c1d145
# @Site    : http://fgyong.cn 兜兜转转的技术博客
# @File    : upload.py
# @Software: PyCharm
# 使用本脚本需要将本脚本移动至flutter项目根项目层级
#
import requests
import os
import getopt
import sys
from enum import Enum
from requests_toolbelt.multipart import encoder


#  检查第三方包是否安装
def check_requirement(package):
    try:
        exec("import {0}".format(package))
    except ModuleNotFoundError:
        inquiry = input(
            "This script requires {0}. Do you want to install {0}? [y/n]".format(package))
        while (inquiry != "y") and (inquiry != "n") and (inquiry != "N") and (inquiry != "Y"):
            inquiry = input(
                "This script requires {0}. Do you want to install {0}? [y/n]".format(package))
        if inquiry == "y" or inquiry == "Y":
            import os
            print("Execute commands: pip install {0}".format(package))
            if int(sys.version.split('.')[0]) == 2:
                os.system("pip install {0}".format(package))
            elif int(sys.version.split('.')[0]) == 3:
                os.system("pip3 install {0}".format(package))
        else:
            print("{0} is missing, so the program exits!".format(package))
            exit(-1)


#  编译类型
class BuildMode(Enum):
    all = 0
    debug = 1
    release = 2

    @staticmethod
    def load(value):
        if value == 0:
            return BuildMode.all
        elif value == 1:
            return BuildMode.debug
        elif value == 2:
            return BuildMode.release

#  上传类型
class UploadMode(Enum):
    all = 0
    apk = 1
    ipa = 2

    @staticmethod
    def load(value):
        if value == 0:
            return UploadMode.all
        elif value == 1:
            return UploadMode.apk
        elif value == 2:
            return UploadMode.ipa

#  编译类别 安卓还是ios
class BuildType(Enum):
    all = 0
    apk = 1
    ipa = 2

    @staticmethod
    def load(value):
        if value == 1:
            return BuildType.apk
        elif value == 2:
            return BuildType.ipa
        elif value == 0:
            return BuildType.all


class Runner:
    buildMode = BuildMode.release
    uploadMode = None
    buildType = None
    fileLength = 1
    currentPath = ''

    def main(self, argv):
        try:
            ops, args = getopt.getopt(
                argv, '-h-b:-u:-t:', ['help', 'build=', 'upload=', 'type='])
        except getopt.GetoptError:
            sys.exit()
        count = 0
        for op, value in ops:
            if op in ('-h', '--help'):
                count += 1
                self.pHelp()
                return
            elif op in ('-u', '--upload'):
                count += 1
                if ['all', 'apk', 'ipa'].__contains__(value) == False:
                    self.pHelp()
                else:
                    self.uploadMode = UploadMode(
                        ['all', 'apk', 'ipa'].index(value))
            elif op in ('-b', '--build'):
                count += 1
                # self.p('-b:{0}'.format(value))
                if ['all', 'debug', 'release'].__contains__(value) == False:
                    self.pHelp()
                else:
                    self.buildMode = BuildMode.load(
                        ['all', 'debug', 'release'].index(value))
            elif op in ('-t', '--type'):
                count += 1
                # self.p('-t:{0}'.format(value))
                if ['all', 'apk', 'ipa'].__contains__(value) == False:
                    self.pHelp()
                else:
                    self.buildType = BuildType.load(
                        ['all', 'apk', 'ipa'].index(value))
        if count == 0:
            self.pHelp()
            return
        else:

            if not self.buildType:
                self.p('\n 请输入 -b buildType 参数 \n')
                self.pHelp()
                return
            if not self.uploadMode:
                self.p('\n 请输入 -u uploadMode 参数 \n')
                self.pHelp()
                return

        self.p('{0}{1}{2}'.format(self.buildMode,
               self.buildType, self.uploadMode))
        # 开始编译和上传ipa
        self.currentPath = os.getcwd()
        self.build()
        self.upload()

    def p(self, str):
        # 结束标示： \n\033[0m
        print('\033[0;35;48m  '+str+'')

    def pHelp(self):
        self.p('-h --help 获取帮助信息')
        self.p('-b --build 编译模式[release(default),debug,all]')
        self.p('-u --upload 上传模式[apk,ipa]')
        self.p('-t --type  编译种类[all,apk,ipa]')

    def buildAPK(self):
        try:
            if self.buildMode == BuildMode.release:
                os.system('flutter --no-color build apk'),
            if self.buildMode == BuildMode.debug:
                os.system('flutter --no-color build apk --debug'),
        except KeyError as e:
            print(e)

    def buildIPA(self,):
        try:
            if self.buildMode == BuildMode.release:
                os.system('flutter --no-color build ios --release'),
                os.chdir(self.currentPath+'/ios')
                plus = 'agvtool next-version -all'
                arch = 'xcodebuild -workspace Runner.xcworkspace -scheme Runner archive -archivePath ./build/Runner.xcarchive'
                export = 'xcodebuild -exportArchive -archivePath $PWD/build/Runner.xcarchive -exportOptionsPlist "./exportOptions.plist" -exportPath ./build/'
                os.system(arch)
                os.system(plus)
                os.system(export)

            elif self.buildMode == BuildMode.debug:
                os.system('flutter --no-color build ios --debug'),
        except KeyError as e:
            print(e)

    # 编译函数
    def build(self,):
        if self.buildType == BuildType.apk:
            self.buildAPK()
        elif self.buildType == BuildType.ipa:
            self.buildIPA()
        elif self.buildType == BuildType.all:
            self.buildAPK()
            self.buildIPA()

    def uploadFile(self, filePath, name='runner.ipa'):
        self.p(' 开始上传' + filePath)
        url = 'https://www.pgyer.com/apiv2/app/upload'
        file = open(filePath, 'rb')
        # dataJson = {
        #     '_api_key': '2e8571d626b9a8c8b752e59624481847',
        #     'buildInstallType': '3',
        #     'buildPassword': '',
        # }
        # 没有进度条的上传
        # fileDir = {'file': file}
        # response = requests.post(url, files=fileDir, data=dataJson)
        # jsonStr = response.json()

        e = encoder.MultipartEncoder(
            fields={
                '_api_key': '57a3e5ff5ad3b6411d5a2680f760445a',
                #     'buildInstallType': '3',
                #     'buildPassword': '',
                'file': (name, file, 'application/x-www-form-urlencoded'),
            },
        )
        m = encoder.MultipartEncoderMonitor(e, self.my_callback)
        h = {'Content-Type': m.content_type, "enctype": "multipart/form-data"}
        self.fileLength = os.path.getsize(filePath)
        r = requests.post(url, data=m, headers=h).json()

        if int(dict(r).get('code')) != 0:
            raise Exception('上传失败:{0}'.format(dict(r).get('message')), )
        else:
            self.p(str(r).replace('u\'', '\''))  # 去掉前边的u
            self.p(filePath + '上传成功')

    def my_callback(self, monitor):
        total = 50
        pro = int((monitor.bytes_read * 1.0)/self.fileLength * 1.0 * total)
        unit = 1000 * 1000.0
        sys.stdout.write(
            '\r' + str(pro * '\033[46;34m \033[0m') + str((total - pro) * '\033[40;30m \033[0m') + '[{0}M/{1}M]'.format(round(monitor.bytes_read/unit, 2), round(self.fileLength/unit, 2)))
        sys.stdout.flush()
    # 上传函数

    def upload(self,):
        currentPath = self.currentPath
        if self.uploadMode == UploadMode.apk or self.uploadMode == UploadMode.all:
            if self.buildMode == BuildMode.debug or self.buildMode == BuildMode.all:
                filePath = currentPath+'/build/app/outputs/apk/debug/app-debug.apk'

                try:
                    self.uploadFile(filePath, name='app-debug.apk')

                except Exception as e:
                    self.p(e.__str__())
                    exit()
            if self.buildMode == BuildMode.release or self.buildMode == BuildMode.all:
                filePath = currentPath+'/build/app/outputs/apk/release/app-release.apk'
                try:
                    self.uploadFile(filePath, name='app-release.apk')
                except Exception as e:
                    self.p(e.__str__())
                    exit()

        if self.uploadMode == UploadMode.ipa or self.uploadMode == UploadMode.all:
            f = os.getcwd()+'/ios/build/Runner.ipa'
            if os.path.isdir(f) or os.path.isfile(f):
                self.uploadFile(f, name='Runner.ipa')
                mv = 'mv ios/build/Runner.ipa ios/output/Runner.ipa'
                rm = 'rm -rf ios/build/'
                os.system(mv)  # 移动ipa
                os.system(rm)  # 删除 build文件夹
            else:
                self.p('没有发现ipa文件:'+f)


if __name__ == '__main__':
    libs = ['requests', 'requests_toolbelt']
    for i in libs:
        check_requirement(i)  # 检查是否安装该安装包
    Runner().main(sys.argv[1:])
